### ISSUE TITLE
Change image resolution

### ISSUE THREAD

**Issue by 25benjaminli** [facebookresearch/segment-anything-2#138]
Similar to SAM 1, SAM 2 was trained on 1024x1024 images. I'm wondering whether it's possible to adapt SAM 2 to 512x512 images without resizing those images (since that takes a lot of time). This was possible in the old SAM 1 by resizing the positional embeddings, but I'm not sure how to go about it in this new repository. After changing `image_size` in the config, the only thing that seemed to cause an error was 
```
feats = [
            feat.permute(1, 2, 0).view(1, -1, *feat_size)
            for feat, feat_size in zip(vision_feats[::-1], self._bb_feat_sizes[::-1])
][::-1]
```

I know SAM 1 uses a typical VIT, but SAM 2 uses hiera (which I'm not very familiar with). I changed the features in `SAM2ImagePredictor` from:
```
self._bb_feat_sizes = [
            (256, 256),
            (128, 128),
            (64, 64),
        ]
```
to:
```
self._bb_feat_sizes = [
            (128, 128),
            (64, 64),
            (32, 32),
        ]     
```

and no errors were outputted, but I am not sure if the segmentation is working properly. Would appreciate feedback on this approach or any alternative methods!

**Comment by brisyramshere** [comment]
do you find solution?

**Comment by heyoeyo** [comment]
> do you find solution?

The change mentioned to the feature sizes list along with a corresponding change to the `image_size` setting (inside the model .yaml configs) should work. More generally, I think you can auto calculate the feature size list in the [SAM2ImagePredictor](https://github.com/facebookresearch/segment-anything-2/blob/0230c5ff9335347bef063a2d88b9ede907f1be01/sam2/sam2_image_predictor.py#L59-L63) with something like:
```python
# Spatial dim for backbone feature maps
hires_size = self.model.image_size // 4
self._bb_feat_sizes = [[hires_size // (2**k)]*2 for k in range(3)]
```

This way you only need to change the `image_size` setting inside the model .yaml files (though only certain image sizes will work without other changes).

If you just want to quickly test out how the models perform at different sizes, I have a script ([run_image.py](https://github.com/heyoeyo/muggled_sam?tab=readme-ov-file#run-image)) that can do this interactively using either SAM v1 or v2 models. You can adjust the processing size to 512 by adding a flag `-b 512` when running the script.

**Comment by brisyramshere** [comment]
> > do you find solution?
> 
> The change mentioned to the feature sizes list along with a corresponding change to the `image_size` setting (inside the model .yaml configs) should work. More generally, I think you can auto calculate the feature size list in the [SAM2ImagePredictor](https://github.com/facebookresearch/segment-anything-2/blob/0230c5ff9335347bef063a2d88b9ede907f1be01/sam2/sam2_image_predictor.py#L59-L63) with something like:
> 
> ```python
> # Spatial dim for backbone feature maps
> hires_size = self.model.image_size // 4
> self._bb_feat_sizes = [[hires_size // (2**k)]*2 for k in range(3)]
> ```
> 
> This way you only need to change the `image_size` setting inside the model .yaml files (though only certain image sizes will work without other changes, multiples of 128 should be safe).
> 
> > ...or any alternative methods
> 
> If you just want to quickly test out how the models perform at different sizes, I have a script ([run_image.py](https://github.com/heyoeyo/muggled_sam?tab=readme-ov-file#run-image)) that can do this interactively using either SAM v1 or v2 models. You can adjust the processing size to 512 by adding a flag `-b 512` when running the script.

Thank you very much.
But since the pre-trained model is of 1024 resolution, will the performance decline to a certain extent after changing it to 768 or 512 resolution?

**Comment by heyoeyo** [comment]
> will the performance decline to a certain extent

Yes, the shape of the masks will get more blocky looking at lower resolutions and eventually masking fails entirely (or at least requires more prompts to get a good result). The reverse is also (somewhat) true, higher resolutions (above 1024) give cleaner mask edges, though at some point the masking starts to fail.

Here's a comparison between 256, 512, 1024 and 2048 resolutions using SAMv2-large with a single foreground prompt point (same for all examples):

![sam2_res_comparison](https://github.com/user-attachments/assets/19d2718c-ac41-4eed-9c74-d843a740ab13)
The 4 masks output by the model are shown on the right side of each picture. The 1024 & 512 resolutions are not so different, but you can see (especially in the masks) that the results are more unstable for 256 or 2048.

**Comment by 25benjaminli** [comment]
@heyoeyo Thanks for the info. Your repository is impressive. It does make sense that the segmentations would get more blocky as the images get smaller. This is because, if I remember correctly, the decoder outputs a mask four times smaller in width & height (so upscaling can result in it looking blocky). 

This is a tangentially related question, but have you found segment anything 2 to create better predictions (at least purely based on observation) compared to segment anything 1?

**Comment by heyoeyo** [comment]
> have you found segment anything 2 to create better predictions

The results can be very similar when allowing for switching between mask outputs (there are major differences in terms of which outputs hold 'whole object' vs. 'sub-component' results between v1 and v2). Here's a comparison of v1 large (left) vs. v2 large (right) using the mask outputs that target sub-components:

![v1_vs_v2_large](https://github.com/user-attachments/assets/50337088-450d-4ea3-ba47-3baf4824afa1)

It's worth noting the resource usage for this example:
| Version | Model size (MB) | Speed (ms) | VRAM (MB) |
| --------| -----------|-------|------|
| V1      | 1200      | 200 | 2300 |
| V2      | 900      | 150 | 750 |

Since the resource usage of v2 is so much lower than v1, it's arguably better in general. I would also say v2 base is significantly better than v1 base (better speed, VRAM & mask quality).

The only noticeable advantage of v1 is that is gives more intuitive results when using box prompts, at least compared to the v2 large & small models (base and tiny variants give results similar to v1).

**Comment by Preburk** [comment]
> > do you find solution?
> 
> The change mentioned to the feature sizes list along with a corresponding change to the `image_size` setting (inside the model .yaml configs) should work. More generally, I think you can auto calculate the feature size list in the [SAM2ImagePredictor](https://github.com/facebookresearch/segment-anything-2/blob/0230c5ff9335347bef063a2d88b9ede907f1be01/sam2/sam2_image_predictor.py#L59-L63) with something like:
> 
> ```python
> # Spatial dim for backbone feature maps
> hires_size = self.model.image_size // 4
> self._bb_feat_sizes = [[hires_size // (2**k)]*2 for k in range(3)]
> ```
> 
> This way you only need to change the `image_size` setting inside the model .yaml files (though only certain image sizes will work without other changes, multiples of 128 should be safe).
> 
> > ...or any alternative methods
> 
> If you just want to quickly test out how the models perform at different sizes, I have a script ([run_image.py](https://github.com/heyoeyo/muggled_sam?tab=readme-ov-file#run-image)) that can do this interactively using either SAM v1 or v2 models. You can adjust the processing size to 512 by adding a flag `-b 512` when running the script.

Hey, This does not seem to work for boxes. The coordinates gets offset a bit. Same in your demo when running 2048 as image size.

**Comment by Preburk** [comment]
> > The coordinates gets offset a bit
> 
> Do you have an example of this? I don't remember seeing any offset issues on images I've tried, but I do find the box prompting for the v2 models can sometimes give poor results compared to v1. Here's the result I get using a box prompt with the large model at 2048px resolution: ![box_example](https://private-user-images.githubusercontent.com/32405350/369911002-141f8cd2-46aa-4ed8-a3db-2f683fc0731c.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3ODkwMTksIm5iZiI6MTcyNzc4ODcxOSwicGF0aCI6Ii8zMjQwNTM1MC8zNjk5MTEwMDItMTQxZjhjZDItNDZhYS00ZWQ4LWEzZGItMmY2ODNmYzA3MzFjLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAxVDEzMTgzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhYzIwYTNjNTcwOGRlNTI4N2M5YzQxNWIwMjIwOGM4YTM0NjdlMGYzNDFmZDgxZTFhZGI3Y2VmYTgyM2FmZTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-SF6G4S1capnHHZ3hWteofB9C-NhQ21qnf0NK02hMOs)

Here you can see some examples: 
![default](https://github.com/user-attachments/assets/b34462a9-574e-4fc3-be6c-d830119c6984)
![Screenshot from 2024-09-24 15-51-10](https://github.com/user-attachments/assets/a498acd5-e533-48f5-87c2-f9ecb5ea7178)
![Screenshot from 2024-09-24 15-52-05](https://github.com/user-attachments/assets/540046bc-fe41-4cf3-a6e4-980668d5ac7d)

As you can see if one use -b 2048, -b 4096,  the mask gets elsewhere.

**Comment by heyoeyo** [comment]
Oh I see what you mean. I think the appearance of there being an offset is more of a coincidence though. For example, it seems to be generating a mask of the area _around_ the selected hole, but then only highlighting the largest part of that mask, which ends up being in one of the corners and that gives it the effect of looking offset. You can try turning off the `Largest Only` setting (by clicking the button or pressing `tab`) to see what the full mask does when drawing the box.

As far as solutions for this, one option is to try a different model. The base model has the highest resolution position encodings (for some reason), which may make it better at dealing with up-scaled inputs:

![example_base_result](https://github.com/user-attachments/assets/57dbd3cd-c1fb-40ca-ac67-5b1762c7841c)


Alternatively, the SAMv1 models tend to handle up/downscaling much better than the v2 models. They also (generally) give better box prompt results. The downside is that they use a lot more VRAM, though you can try adding the `-ar` flag to that same script to run at the image aspect ratio, which the v1 models handle fairly well. You can also save some memory by using the large v1 model (instead of the huge model) which is usually good enough.

If you want to stick with v2, another possible solution is to pre-crop the image around the area you're trying to segment. This way you can stay at 1024px (which is good for v2) but allocate more of the space to the segmented object. You can add `--crop` when running that script to do this (and I just pushed an update that saves a full-sized copy of the mask, which may be helpful here). Here's an example for that hole:

https://github.com/user-attachments/assets/229f1b3f-4df5-47b1-9c36-97dd23db1533

In this example, the mask after cropping is only 64x64 tokens, but the cropping helps to improve the mask resolution as if a much higher input resolution was used. To be fair, I still found it hard to get consistent results with the large v2 model, even with cropping, it seems to struggle with this kind of image.

**Comment by Preburk** [comment]
> Oh I see what you mean. I think the appearance of there being an offset is more of a coincidence though. For example, it seems to be generating a mask of the area _around_ the selected hole, but then only highlighting the largest part of that mask, which ends up being in one of the corners and that gives it the effect of looking offset. You can try turning off the `Largest Only` setting (by clicking the button or pressing `tab`) to see what the full mask does when drawing the box.
> 
> As far as solutions for this, one option is to try a different model. The base model has the highest resolution position encodings (for some reason), which may make it better at dealing with up-scaled inputs:
> 
> ![example_base_result](https://private-user-images.githubusercontent.com/32405350/372657911-57dbd3cd-c1fb-40ca-ac67-5b1762c7841c.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc4Nzk0NTYsIm5iZiI6MTcyNzg3OTE1NiwicGF0aCI6Ii8zMjQwNTM1MC8zNzI2NTc5MTEtNTdkYmQzY2QtYzFmYi00MGNhLWFjNjctNWIxNzYyYzc4NDFjLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAyVDE0MjU1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM2OTI2OTgyN2Y1OWNiNDJmMWU4ZjYxNzhlY2NlYWQxMzM1MTRjNDg2M2Y1ZDI0MTI5YzNkODQ5MGY4N2QyNjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dcNs9q37CLvQbgWOEH07dhm7bIp8Qny1TuQjC-jNXUA)
> 
> Alternatively, the SAMv1 models tend to handle up/downscaling much better than the v2 models. They also (generally) give better box prompt results. The downside is that they use a lot more VRAM, though you can try adding the `-ar` flag to that same script to run at the image aspect ratio, which the v1 models handle fairly well. You can also save some memory by using the large v1 model (instead of the huge model) which is usually good enough.
> 
> If you want to stick with v2, another possible solution is to pre-crop the image around the area you're trying to segment. This way you can stay at 1024px (which is good for v2) but allocate more of the space to the segmented object. You can add `--crop` when running that script to do this (and I just pushed an update that saves a full-sized copy of the mask, which may be helpful here). Here's an example for that hole:
> precrop_video_example.mp4
> 
> In this example, the mask after cropping is only 64x64 tokens, but the cropping helps to improve the mask resolution as if a much higher input resolution was used. To be fair, I still found it hard to get consistent results with the large v2 model, even with cropping, it seems to struggle with this kind of image.

Thank you for the thorough answer.

I have been testing with using tiles/crops for samv1 and sam-hq previously but they are bit too slow and large for my purposes and v2 set image is still too slow for several crops and batching seem to take an enormous amount of memory even with very small crops(256x256), almost 8gb for 10 crops for the smallest model.

The Base+ v2.1 model works much better and seem to be almost on par with v1, a shame that the other models have lower resolution position encodings which results in the issues mentioned, it would be nice to be able to get higher quality masks without having to resort to cropping/tiling. 

This is probably an easy fix if they retrain the models in the future since they already have done so with the b+ model. Unfortunely i do not have the compute capacity to try and re-train the models myself.

I will try use the base+ in my pipeline on the full-res images and see if the results are satisfactory, otherwise i will have to find an alternative solution.

**Comment by bhack** [comment]
> it would be nice to be able to get higher quality masks without having to resort to cropping/tiling.

is there also an intrinsic limit in the decoder structure with its "relative" aggressive upscale factor if you want to work with highers images/masks?

https://github.com/facebookresearch/sam2/blob/52198ead0eb13ae8270bea6ca768ef175f5bf167/sam2/modeling/sam/mask_decoder.py#L65-L90

**Comment by heyoeyo** [comment]
> batching seem to take an enormous amount of memory... almost 8gb for 10 crops

This seems quite high! Using a batch size of 10 takes ~4GB for the v2 tiny model for me. It might be worth double checking that (1) the batching is being done inside a `torch.inference_mode()` block, so that gradients aren't being recorded and (2) that bfloat16 data type is being used (if possible), since that should cut the VRAM usage in half in case float32 is being used.

Also the model will scale inputs to 1024px regardless of the input size (e.g. 256px crops will be scaled to 1024px). You can change the `image_size` (in the .yaml files) config parameter to have the model work at lower resolutions, which further drops the VRAM down to ~1GB for me.

> is there also an intrinsic limit in the decoder structure

The decoder upscaling is relative, so if a higher processing resolution is used, like 2048px, it will be working from a higher token resolution (e.g. scaling 128x128 up to 512x512). The v2 models also add in higher resolution image features while [performing the upscaling](https://github.com/facebookresearch/sam2/blob/52198ead0eb13ae8270bea6ca768ef175f5bf167/sam2/modeling/sam/mask_decoder.py#L222-L225), so there's some potential for it to be a bit smarter than just a weighted upscaling. It's definitely a trade-off between resource use and mask accuracy, but I wouldn't expect it to be a limiting factor on it's own.

**Comment by bhack** [comment]
> to work from a higher token resolution (e.g. scaling 128x128 up to 512x512)

I think it is hard to learn high Res details if your decoder in the learning stages it is limited at a so low resolution or not?

It is true that out Binary masks could be interpolated without introducing significant artifacts but you are going to recover fine details, small holes and everything else you have available at high res especially now that we have an official training code.

**Comment by heyoeyo** [comment]
> I think it is hard to learn high Res details if your decoder in the learning stages it is limited at a so low resolution or not?

Ya I agree. The 64x64 internal resolution probably has some negative impact on the final mask resolution, but this seems like a required trade-off to avoid higher resource use during inference/training. Though I think the fact that the decoder doesn't do the full 16x upscale (only 4x) may have a bigger negative impact on the end result.

Here's an example with a processing resolution of 256px which leads to having only 16 tokens across:

![token_to_mask_example](https://github.com/user-attachments/assets/2557c334-bbd1-45ae-8964-5a1c93d954bf)

On the left is an input image & box prompt, on the right is the binary mask with a token grid overlay. The grid is meant to indicate the 'positioning' of the image tokens used by the upscaler.

If the low internal resolution (16x16 here) was a limitation, I think we'd expect to see a mask that 'snaps' to the token grid somehow. Instead, it looks like the upscaler can create complex mask shapes & small holes from whatever information is held in the tokens (see the gap between the legs in the middle for example) and seems mostly independent of the grid.

After the decoder upscaler, the result is limited like you say (can't recover fine details/small holes). Here's the actual 256px image & 64px upscaler mask for comparison:

![input_to_upscaler_output_comparison](https://github.com/user-attachments/assets/3f3ee52a-8eaf-4c8b-ab09-a8cc64143bb9)

So any details too small to resolve in the small mask won't appear after (regular) interpolation back to 256px. However the fine detail may be available in the tokens if the decoder upscaler was used for the full 16x scaling instead.

**Comment by bhack** [comment]
> So any details too small to resolve in the small mask won't appear after (regular) interpolation back to 256px. However the fine detail may be available in the tokens if the decoder upscaler was used for the full 16x scaling instead.

Yes, exactly the point is really how much interesting is to increase the input resolution if at the end we have just a big "nearest" interpolation on a binary mask to go back to the original resolution?

I think partially this was caused, other then for resources, by the point that the original authors are not interested to gain more details other then a big nearest upscaling step as there is an intrinsic limit on the annotation quality of the dataset when they are going to put out paper metrics on the table.

As we have the training now with 2.1  it would be nice to have a configurable decoder, still compatible with the current checkpoints, but that it is able at same resolution to get more details on training (e.g. suppose in medical imaging etc..)

**Comment by bhack** [comment]
P.s. You can try e.g. to test the decoder with some selected images of [BIG dataset](https://paperswithcode.com/dataset/big)

**Comment by Preburk** [comment]
> > batching seem to take an enormous amount of memory... almost 8gb for 10 crops
> 
> This seems quite high! Using a batch size of 10 takes ~4GB for the v2 tiny model for me. It might be worth double checking that (1) the batching is being done inside a `torch.inference_mode()` block, so that gradients aren't being recorded and (2) that bfloat16 data type is being used (if possible), since that should cut the VRAM usage in half in case float32 is being used.
> 
> Also the model will scale inputs to 1024px regardless of the input size (e.g. 256px crops will be scaled to 1024px). You can change the `image_size` (in the .yaml files) config parameter to have the model work at lower resolutions, which further drops the VRAM down to ~1GB for me.


Runs in inference mode block and bfloat16 still very high usage, almost no difference regardless which model is used. 
I tried running with 256 as image size in the yaml but then the resulting masks becomes pretty much useless and super low resolution. e.g running your run_image.py with the -b 256 parameter on a 256x256 crop.

I also tried running a 20mp image with base+ model and 4096. The results are ok ~70% of the time but it still struggles with the rest. SAM-HQ still seem to be king if one is looking for quality and predictability, but the tradeoff is speed and resource usage. I guess SAM2 is more designed towards real time inference for videos while sam1 and sam-hq focused on quality image masking. 

Lemme know if you have any suggestion how one can achieve masking on par with the older networks quality wise.

**Comment by heyoeyo** [comment]
> The results are ok ~70% of the time... I guess SAM2 is more designed towards real time inference for videos

I'd agree, it seems the v2 release is really about supporting video instead of doubling down on image segmentation.

> Lemme know if you have any suggestion how one can achieve masking on par with the older networks quality wise.

In terms of reliability/consistency, the best bet might just be to wait for fine tuned variants unfortunately. However, for cropping, 512px seems to work more consistently than 256 for the v2 models, so might be worth trying (also points prompts sometimes stop working at lower resolutions, but box prompts can still work, especially for the non-base models).

**Comment by 25benjaminli** [comment]
@heyoeyo is there code that specifically resizes positional encodings for SAM2? Just curious because I don't fully understand why the weights can be loaded without error (which happens for SAM1). I think it's probably something to do with the new Hiera image encoder

**Comment by heyoeyo** [comment]
> is there code that specifically resizes positional encodings for SAM2

Yes you're right, the Hiera model of SAMv2 already includes code for resizing the position encodings:
https://github.com/facebookresearch/sam2/blob/e22521832f4e8cd1881f6be4a9e48dfea1856430/sam2/modeling/backbones/hieradet.py#L276
It also has an additive tiled component which is unique to the Hiera model afaik (see the paper: [Window Attention is Bugged](https://arxiv.org/abs/2311.05613)).

SAMv1 can do the same thing (not including the tiling part), but it's not included in the original code release for some reason.

**Comment by urbste** [comment]
Thanks for the discussion! I made some tests myself and evaluated with box prompts on the Coco 2017 validation set. You can find the results [here](https://github.com/urbste/nanosam2?tab=readme-ov-file#testing-sam21-hiera-backbones-with-different-resolutions). TL;DR on Coco the mIoU drops from 75.0 to 68.4 if you go from 1024 to 512. So quite significant.
